marsha/core/tests/test_models.py

```diff
@@ -0,0 +1,49 @@
+"""Tests for the models in the ``core`` app of the Marsha project."""
+from typing import Type
+
+from django.db import IntegrityError
+from django.test import TestCase
+
+from . import factories
+
+
+# We don't enforce argument documentation in tests
+# pylint: disable=missing-param-doc,missing-type-doc
+
+
+class SoftDeletionTestCase(TestCase):
+    """Will test soft-deletion of our different models."""
+
+    def _test_soft_deletion(
+        self, factory: Type[factories.BaseSoftDeleteFactory], **kwargs
+    ):
+        """Ensure that soft deletion works for the model represented by the factory."""
+        model = factory._meta.model  # noqa
+
+        obj = factory.create(**kwargs)
+        self.assertIsNone(obj.deleted)
+
+        obj.delete()
+        self.assertIsNotNone(obj.deleted)
+
+        # object should be invisible to normal queryset
+        with self.assertRaises(model.DoesNotExist):
+            model.objects.get(pk=obj.pk)
+
+        # but still present in deleted mode
+        self.assertEqual(model.objects.deleted_only().get(pk=obj.pk), obj)
+
+    def test_user_is_soft_deleted(self):
+        """Ensure soft deletion work as expected for users."""
+        self._test_soft_deletion(factories.UserFactory)
+
+    def test_username_is_unique_even_if_deleted(self):
+        """Ensure that a username of a deleted user cannot be used by another."""
+        user = factories.UserFactory.create(username="foo")
```

Comment:
why set username. It's the goal of the factory to do this. You can have a field in UserFactory:
```python
username = factory.Faker('username')
```

Reply:
It's for uniqueness, I force it to a specific value to use the same later to try to create another user with the same username.